### PR TITLE
ColBERT Passage Ranking 

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by jobergum** [comment]
A little bit to review for the weekend @kkraune :)

CC @lesters @thigm85

File: msmarco-ranking/document-ranking.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  **NDCG@10 0.6136**
+
+# Run Time Serving Performance
+Vespa's evaluation of GBDT models is hyper optimized after 20 years of using GBDT at scale so end to end serving time is roughly 20 ms. 
+Vespa supports using multiple threads per *query* 
+and in our experiment we use up to 12 threads per query. 
+This allows scaling latency per node and make use of multi-core cpu architectures efficiently.
+
+Below the top two documents
+ranked for the question *when was nelson mandela born*.
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:143]:
```suggestion
See the top two documents ranked for the question *when was nelson mandela born* below.
```

File: msmarco-ranking/document-ranking.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+* Git client to checkout the sample application repository
+* Java 11, Maven and python3 installed
+* Operating system: macOS or Linux, Architecture: x86_64
+
+See also [Vespa quick start guide](https://docs.vespa.ai/documentation/vespa-quick-start.html).
+
+First, we retrieve the sample app:
+
+<pre data-test="exec">
+$ git clone https://github.com/vespa-engine/sample-apps.git
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:175]:
```suggestion
$ git clone --depth 1 https://github.com/vespa-engine/sample-apps.git
```

File: msmarco-ranking/document-ranking.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+now, wait for the application to start. 
+
+<pre data-test="exec" data-test-wait-for="200 OK">
+$ curl -s --head http://localhost:8080/ApplicationStatus
+</pre>
+
+## Feeding Sample Data 
+
+Feed sample data. We feed the documents using the [Vespa http feeder
+client](https://docs.vespa.ai/documentation/vespa-http-client.html):
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:217]:
```suggestion
client](https://docs.vespa.ai/en/vespa-http-client.html):
```

File: msmarco-ranking/document-ranking.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+now, wait for the application to start. 
+
+<pre data-test="exec" data-test-wait-for="200 OK">
+$ curl -s --head http://localhost:8080/ApplicationStatus
+</pre>
+
+## Feeding Sample Data 
+
+Feed sample data. We feed the documents using the [Vespa http feeder
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:216]:
```suggestion
Feed sample data using the [Vespa http feeder
```

File: msmarco-ranking/document-ranking.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+$ docker exec vespa bash -c 'java -jar /opt/vespa/lib/jars/vespa-http-client-jar-with-dependencies.jar \
+    --file /MSMARCO/sample-feed/sample_regular_fields.jsonl --host localhost --port 8080'
+</pre>
+
+<pre data-test="exec">
+$ docker exec vespa bash -c 'java -jar /opt/vespa/lib/jars/vespa-http-client-jar-with-dependencies.jar \
+    --file /MSMARCO/sample-feed/sample_doc_t5_query.jsonl --host localhost --port 8080'
+</pre>
+
+Now all the data is in place and one can play around with the search interface (Though only searching 1K documents)
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:229]:
```suggestion
Now all the data is in place and one can play around with the query interface (Though only searching 1K documents)
```

File: msmarco-ranking/document-ranking.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+<pre>
+$ curl -s http://localhost:8080/document/v1/msmarco/doc/docid/D1840066 |python -m json.tool
+</pre>
+
+Do a query 
+<pre>
+curl -s "http://localhost:8080/search/?query=what%20is%20the%20definition%20of%20business%20law?&ranking=ltr" |python -m json.tool
+</pre>
+
+The data set is small, but one gets a feel for how the data and how the doc tttt query expansion work. Note that negative relevance scores from the GBDT evaluation
```
Comment (kkraune) [inline msmarco-ranking/document-ranking.md:241]:
```suggestion
The data set is small, but one gets a feel for how the data and how the doc ttttt query expansion work. Note that negative relevance scores from the GBDT evaluation
```

not sure if a t is missing here?

File: msmarco-ranking/sample-feed/query.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "yql": "select id, text from sources passage where userQuery();",
+  "ranking": "colbert",
+  "query": "what was the manhattan project?",
+  "hits": 5,
+  "searchChain":"passageranking",
+  "timeout":'2s'
```
Comment (kkraune) [inline msmarco-ranking/sample-feed/query.json:7]:
```suggestion
  "timeout": "2s"
```

File: msmarco-ranking/src/main/python/train.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 # Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#This is an example of how to train a LTR model using Lighgbm
```
Comment (kkraune) [inline msmarco-ranking/src/main/python/train.py:3]:
```suggestion
#This is an example of how to train a LTR model using LightGBM
```

File: msmarco-ranking/src/test/ai/vespa/searcher/RetrievalSearcherTest.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package ai.vespa.searcher;
```
Comment (kkraune) [inline msmarco-ranking/src/test/ai/vespa/searcher/RetrievalSearcherTest.java:1]:
```suggestion
// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

package ai.vespa.searcher;
```

**Pull request merged**